Cutelyst/application.cpp

```diff
@@ -441,6 +442,96 @@ bool Application::enginePostFork()
     return true;
 }
 
+void Application::addTranslator(const QLocale &locale, QTranslator *translator)
+{
+    Q_D(Application);
+    Q_ASSERT_X(translator, "add translator to application", "invalid QTranslator object");
+    if (d->translators.contains(locale)) {
+        QVector<QTranslator*> trs = d->translators.value(locale);
+        trs.prepend(translator);
+        d->translators.insert(locale, trs);
+    } else {
+        d->translators.insert(locale, QVector<QTranslator*>(1, translator));
+    }
+}
+
+void Application::addTranslator(const QString &locale, QTranslator *translator)
+{
+    addTranslator(QLocale(locale), translator);
+}
+
+void Application::addTranslators(const QLocale &locale, const QVector<QTranslator *> &translators)
+{
+    Q_D(Application);
+    Q_ASSERT_X(!translators.empty(), "add translators to application", "empty translators vector");
+    if (d->translators.contains(locale)) {
+        QVector<QTranslator*> trs = d->translators.value(locale);
+        for (auto it = translators.crbegin(); it != translators.crend(); ++it) {
+            trs.prepend(*it);
+        }
+        d->translators.insert(locale, trs);
+    } else {
+        d->translators.insert(locale, translators);
+    }
+}
+
+static void replacePercentN(QString *result, int n)
```

Comment:
Yes, I think it is needed, because we can not use QCoreApplication::translate(). QCoreApplication implements this function internally (it is copy pasted from there, that's why indentation is different) and uses it in a similar way. [See Qt sources](http://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/kernel/qcoreapplication.cpp#n1982). Grantlee also reimplements this function. [See Grantlee sources](https://github.com/steveire/grantlee/blob/master/templates/lib/qtlocalizer.cpp#L92). So, substitution is not done in QTranslator::translate() but in QCoreApplication::translate() (what is called by QObject::tr() etc).